test/libsolidity/util/ContractABIUtils.cpp

```diff
@@ -33,55 +39,256 @@ using namespace dev::solidity::test;
+}
+
+auto isIntArray(string _type) -> bool
+{
+	return regex_match(_type, regex{"(int\\d*)(\\[\\d+\\])"});
+}
+
+auto isStringArray(string _type) -> bool
+{
+	return regex_match(_type, regex{"(string)(\\[\\d+\\])"});
+}
+
+auto isTuple(string _type) -> bool
+{
+	return regex_match(_type, regex{"(tuple)"});
+}
+
+auto isTupleArray(string _type) -> bool
+{
+	return regex_match(_type, regex{"(tuple)(\\[\\d+\\])"});
+}
+
+auto functionMatchesSignature(Json::Value const& _contractABIFunction, string const& _signature) -> bool
+{
+	string functionName{_signature.substr(0, _signature.find("("))};
+	string functionParameters{_signature.substr(_signature.find("(") + 1, _signature.find(")") - 2)};
+
+	vector<string> parameterTypes;
+	if (!functionParameters.empty())
+		boost::split(parameterTypes, functionParameters, [](char c){return c == ',';});
+
+	vector<string> abiParameterTypes;
+	for (auto const& input: _contractABIFunction["inputs"])
+		abiParameterTypes.push_back(input["type"].asString());
+
+	return (
+		_contractABIFunction["name"] == functionName &&
+		parameterTypes == abiParameterTypes
+	);
+}
+
+}
+
+boost::optional<dev::solidity::test::ParameterList> ContractABIUtils::parametersFromJson(
+	ErrorReporter& _errorReporter,
 	Json::Value const& _contractABI,
-	string const& _functionName
-) const
+	string const& _functionSignature
+)
 {
-	ParameterList abiParams;
+	ParameterList addressTypeParams;
+	ParameterList valueTypeParams;
+	ParameterList dynamicTypeParams;
+
+	ParameterList finalParams;
+
+	if (!_contractABI)
+		return boost::none;
+
 	for (auto const& function: _contractABI)
-		if (function["name"] == _functionName)
+	{
+		if (functionMatchesSignature(function, _functionSignature))
 			for (auto const& output: function["outputs"])
 			{
-				auto types = fromTypeName(output["type"].asString());
-				for (auto const& type: types)
-					abiParams.push_back(Parameter{bytes(), "", type, FormatInfo{}});
+				string type = output["type"].asString();
+
+				ABITypes addressTypes;
+				ABITypes valueTypes;
+				ABITypes dynamicTypes;
+
+				if (appendTypesFromName(output, addressTypes, valueTypes, dynamicTypes))
+				{
+					for (auto const& type: addressTypes)
+						addressTypeParams.push_back(Parameter{bytes(), "", type, FormatInfo{}});
+					for (auto const& type: valueTypes)
+						valueTypeParams.push_back(Parameter{bytes(), "", type, FormatInfo{}});
+					for (auto const& type: dynamicTypes)
+						dynamicTypeParams.push_back(Parameter{bytes(), "", type, FormatInfo{}});
+				}
+				else
+				{
+					_errorReporter.warning(
+						"Could not convert \"" + type +
+						"\" to internal ABI type representation. Falling back to default encoding."
+					);
+					return boost::none;
+				}
+
+				finalParams += addressTypeParams + valueTypeParams;
+
+				addressTypeParams.clear();
+				valueTypeParams.clear();
 			}
+	}
 
-	return abiParams;
+	return boost::optional<ParameterList>(finalParams + dynamicTypeParams);
```

Comment:
This should return `boost::none` and the actual return should be in the loop iteration that found the function